### PR TITLE
SQL: Implement filter push-down for GROUP BYs

### LINKED ISSUE #69758: SQL: Verifier confuses Filter (WHERE) on top of Aggregate with HAVING
The following queries do not pass the Verifier:

```java
    public void testGroupByWhereSubselect() {
        accept("SELECT b, a FROM (SELECT bool as b, AVG(int) as a FROM test GROUP BY bool) WHERE b = false");
        accept("SELECT b, a FROM (SELECT bool as b, AVG(int) as a FROM test GROUP BY bool HAVING AVG(int) > 2) WHERE b = false");
    }
```

rather they give the following `VerificationException`:

```
Cannot use HAVING filter on non-aggregate [bool]; use WHERE instead
```

even though the query is valid and filter could be pushed down below the `Aggregate` and can be flattened into a single query.

### PR DESCRIPTION
**Pull request by palesz**
Previously we were not able to execute queries like the one below,
where we filter on the groupings of the `Aggregate`s. The reasons are
the limitations of the `bucket_selector`s `bucket_path` [0] that cannot
filter on non-numerical keys.

The implemented optimization transforms the following query:

```sql
SELECT i, a
FROM (
     SELECT int as i, AVG(float) as a
     FROM test
     GROUP BY int
) WHERE i = 1 AND a > 1
```

To one equivalent to this:

```sql
SELECT i, a
FROM (
     SELECT int as i, AVG(float) as a
     FROM test
     WHERE int = 1
     GROUP BY int
) WHERE a > 1
```

The latter query can be translated to Query DSL.

The change also makes it possible to use a query with HAVING in a
subselect with filters in the outer query. We can have multiple levels
of queries on a single GROUP BY with filters on multiple levels. At
the end the filters will be collapsed, the conditions on the
aggregates will transformed into an equivalent of HAVING, while the
conditions on groupings will be pushed below the GROUP BY query.

Note: because of the limitation [0] mentioned above
(sub-)conditions that use expressions both on groupings and aggregates
of the GROUP BY and unsplittable (groupings and aggregates are not in
separate child of a conjunction) cannot be handled.

For example the following query will fail with a VerifierException:

```sql
SELECT *
FROM (
     SELECT int, count(*) as cnt
     FROM test
     GROUP BY int
)
WHERE int > 2 OR cnt > 1
```

because the `int > 2 OR cnt > 1` filters both on the groupings (`int >
2`) and the aggregates (`cnt > 1` that translates to `count(*) > 1`).

Note: Queries like

```sql
SELECT MAX(int) FROM test WHERE MAX(int) > 10
```

are still not allowed (HAVING should be used).

Also fixes #69758, by making the queries work utilizing the filter
push-down, but note that there are changes in the
`EXPLAIN (PLAN ANALYZED)` outputs (`SubqueryAlias` node appears).

[0] https://github.com/elastic/elasticsearch/issues/23874

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 1]
```diff
@@ -124,7 +124,6 @@ public Analyzer(Configuration configuration, FunctionRegistry functionRegistry,
                 //new ImplicitCasting()
                 );
         Batch finish = new Batch("Finish Analysis",
-                new PruneSubqueryAliases(),
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:127]: Shouldn't we remove this inner class? is it used anywhere else?

Comment (palesz) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:127]: Still required. it was moved to the Optimizer.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 2]
```diff
@@ -2697,8 +2697,49 @@ public void testSubqueryGroupByFilterAndOrderByByRealiased() {
         optimizeAndPlan("SELECT g as h FROM (SELECT date AS f, int AS g FROM test) WHERE h IS NOT NULL GROUP BY h ORDER BY h ASC");
     }
     
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/69758")
-    public void testFilterAfterGroupBy() {
-        optimizeAndPlan("SELECT j AS k FROM (SELECT i AS j FROM ( SELECT int AS i FROM test) GROUP BY j) WHERE j < 5");
+    public void testFilterOnAggregatesSameAsHaving() {
+        EsQueryExec havingPlan =
```
Comment (matriv) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:2701]: For this and the other tests, it would be nice to also have a simple query (without subselect) and check that the plans are equivalent.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java [hunk 3]
```diff
@@ -113,7 +113,7 @@ PhysicalPlan fold(PhysicalPlan plan) {
     }
 
     @Override
-    protected Iterable<RuleExecutor<PhysicalPlan>.Batch> batches() {
+    protected Iterable<RuleExecutor<PhysicalPlan>.Batch> batches() {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java:116]: Unnecessary whitespace.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 4]
```diff
@@ -482,6 +508,217 @@ protected LogicalPlan rule(OrderBy ob) {
...
+                    // only add, if the issues did not propagate from the children
+                    if (e.children().stream().noneMatch(expressionsWithIssues::contains)) {
+                        expressionsToReport.add(e);
+                    }
+                }
+            });
+            
+            expressionsToReport.forEach(e -> 
+                failures.add(fail(e, "[{}] is part of a condition on top of a GROUP BY that references both a grouping " +
+                "and an aggregate in an unsplittable expression, cannot be translated (yet)", e)));
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:626]: I'd say to remove the ` (yet)` as this always gives a promise to the user.

**Review by matriv — COMMENTED** [review]
Did a quick first pass, that's a nice improvement on the subselects support.
Left a first round of comments.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java (hunk 2 above)
Comment (palesz) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:2701]: Not sure what do you mean by "simple query". Can you give me an example?

Comment (matriv) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:2701]: `SELECT bool as b, AVG(int) as a FROM test GROUP BY bool HAVING AVG(int) > 2`

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 5]
```diff
@@ -2697,8 +2697,49 @@ public void testSubqueryGroupByFilterAndOrderByByRealiased() {
...
+    public void testMultiProjectFilterOnAggregate() {
+        EsQueryExec manual = (EsQueryExec) 
+            optimizeAndPlan("SELECT int as l FROM test WHERE int < 5 AND int > 3 GROUP BY int HAVING COUNT(int) > 1");
+        EsQueryExec optimized = (EsQueryExec) optimizeAndPlan("SELECT k as l FROM (SELECT j AS k FROM (SELECT i AS j FROM " +
+            "( SELECT int AS i FROM test) GROUP BY j HAVING COUNT(i) > 1) WHERE j < 5) WHERE k > 3");
+        assertEquals(manual.queryContainer().toString(), optimized.queryContainer().toString());
+    } 
+
+    public void testSubqueryWithAlias() {
+        optimizeAndPlan("SELECT t.int FROM test AS t LIMIT 1");
```
[the remark this answers is not in the document]
Comment (matriv) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:2772]: Maybe I'm confused, but then the name of the test method is not correct: `SubqueryWithAlias`, This is an alias for the relation `test`. Could you also add some assertion on the resulting plan?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 6]
```diff
@@ -484,6 +509,247 @@ protected LogicalPlan rule(OrderBy ob) {
...
+            
+            return conditionOnAggregates == null ? newChild : new Filter(plan.source(), newChild, conditionOnAggregates);
+        }
+
+        public static void verify(
+            Filter plan, Set<Failure> failures, AttributeMap<Expression> attributeRefs) {
+            Aggregate aggregate = findFirstAggregateChild(plan);
+            java.util.function.Function<Expression, Expression> resolver = e -> attributeRefs.resolve(e, e);
+            Expression condition = plan.condition().transformDown(resolver);
+            if (aggregate != null) {
```
Comment (bpintea) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:638]: Should this check be moved above?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 7]
```diff
@@ -484,6 +509,247 @@ protected LogicalPlan rule(OrderBy ob) {
...
+                    Filter filterOnGroupings = new Filter(Source.EMPTY, aggregate.child(), conditionOnGroupings);
+                    return new Aggregate(aggregate.source(), filterOnGroupings, aggregate.groupings(), aggregate.aggregates());
+                }
+                return p;
+            });
+            
+            return conditionOnAggregates == null ? newChild : new Filter(plan.source(), newChild, conditionOnAggregates);
+        }
+
+        public static void verify(
```
Comment (bpintea) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:633]: This function seems only to only be called from the Verifier: would it make sense to have it here?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 8]
```diff
@@ -484,6 +509,247 @@ protected LogicalPlan rule(OrderBy ob) {
...
+            // we have an aggregate below the filter, so let's try to break up the filter into two separate ones
+            // let's try to break up the condition into two parts
+            final Expression condition = plan.condition();
+
+            final IdentityHashMap<Expression, Markers> markerMap = 
+                markConditionChildren(aggregate, condition, java.util.function.Function.identity());
+            Set<Failure> failures = new LinkedHashSet<>();
+            verifyCanPushDown(condition, markerMap, failures);
+            if (failures.isEmpty() == false) {
+                throw new QlIllegalArgumentException(Failure.failMessage(failures));
```
Comment (bpintea) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:567]: Does this mean that we could optimise a query, but then still fail it?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 9]
```diff
@@ -484,6 +509,247 @@ protected LogicalPlan rule(OrderBy ob) {
...
+        PushDownFilters() {
+            super(TransformDirection.UP);
+        }
+
+        @Override
+        protected LogicalPlan rule(Filter plan) {
+            // Filter[condition=f(grouping, aggregate)] -> ... -> Aggregate structure will be broken up into
+            // two separate filters if possible, so the filter on the grouping can be pushed down
+            // SELECT groupings, aggregates, ... FROM ... GROUP BY [groupings]
+            // the aggregates in this context mean only the aggregates that are not part of the grouping
```
Comment (bpintea) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:548]: Quoting might improve a bit the legibility.
```suggestion
            // the "aggregates" in this context mean only the aggregates that are not part of the grouping
```

**Comment by costin** [comment]
Superseded by #71279 and #71362
